substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -568,10 +559,16 @@ impl<T: Config> Pallet<T> {
 					.checked_add(&amount)
 					.ok_or(ArithmeticError::Overflow)?
 			} else {
+				// if this is the first time delegation, increment provider count. This ensures that
+				// amount including existential deposit can be held.
+				frame_system::Pallet::<T>::inc_providers(&delegator);
 				amount
 			};
 
-		Delegation::<T>::new(&agent, new_delegation_amount).update_or_kill(&delegator);
+		// try to hold the funds.
+		T::Currency::hold(&HoldReason::StakingDelegation.into(), &delegator, amount)?;
+
+		let _ = Delegation::<T>::new(&agent, new_delegation_amount).update(&delegator);
```

Comment:
I find `::new().update()` a bit counter intuitive, as you have already 'read' the corresponding delegator from storage above. The root cause is that you read the target delegator, but you drop it and only keep `new_delegation_amount`. Here's an alternative:

```rust
if let Some(mut existing_delegation) = Delegation::<T>::get(&delegator) {
    ensure!(existing_delegation.agent == agent, Error::<T>::InvalidDelegation);
    existing_delegation
        .amount
        .checked_add(&amount)
        .ok_or(ArithmeticError::Overflow)?;
    existing_delegation
} else {
    // if this is the first time delegation, increment provider count. This ensures that
    // amount including existential deposit can be held.
    frame_system::Pallet::<T>::inc_providers(&delegator);
    Delegation::<T>::new(&agent, amount)
}
.update(&delegator);
```

Reply:
Makes sense 👍